[â† Back to Index](../project-policy-index.md)

---

# Section 1: Fundamentals

**Contains:** Introduction, Core Principles, AI Agent Automation Rules, PRD Alignment, Change Management

---

## 1. Introduction

This policy provides a **single, authoritative, machine-readable source of truth** for AI coding agents and humans. It ensures all work is governed by clear, unambiguous rules and workflows.

### 1.1 Actors

| Actor | Role | Responsibilities |
|-------|------|------------------|
| **User** | Decision Maker | Defines requirements, prioritizes work, approves changes, accountable for all code |
| **AI_Agent** | Executor | Implements User's instructions precisely as defined by PBIs and tasks |

### 1.2 Architectural Compliance

âœ… All tasks must be explicitly defined and agreed upon before implementation
âœ… All code changes must be associated with a specific task
âœ… All PBIs must be aligned with the PRD when applicable

---

## 2. Fundamental Principles

### 2.1 Core Principles

These principles are **MANDATORY** and govern all work:

#### ğŸ¯ Task-Driven Development
- âŒ **NO code changes** without an agreed-upon task
- âœ… Every change must have explicit authorization

#### ğŸ“¦ PBI Association
- âŒ **NO task creation** without an associated PBI
- âœ… All work traces back to a Product Backlog Item

#### ğŸ“– PRD Alignment
- âœ… If a PRD exists, all PBI features must align with its scope
- âœ… Sense-check all features against PRD requirements

#### ğŸ‘¤ User Authority
- âœ… User is the **sole decider** for scope and design
- âœ… User retains responsibility for all code changes (even if AI implements)

#### ğŸš« Prohibition of Unapproved Changes
- âŒ **EXPRESSLY PROHIBITED:** Any changes outside explicit task scope
- âœ… Identify scope creep â†’ roll back â†’ create new task

#### ğŸ”„ Task Status Synchronization
- âœ… Status in `1-tasks.md` **MUST** match status in individual task file
- âœ… Update **BOTH** locations immediately on status change

#### ğŸ“„ Controlled File Creation
- âŒ **DO NOT** create files outside defined structures (PBIs, tasks, source code)
- âœ… Get **explicit User confirmation** before creating any standalone files

#### ğŸ“š External Package Research
To avoid hallucinations when using external packages:

1. **Research first:** Use web search to find official documentation
2. **Create guide:** For each package, create `tasks/<task-id>-<package>-guide.md`
3. **Document:** Include:
   - Date stamp
   - Link to original docs
   - API usage examples in project's language
   - Code snippets

**Example:** For task 2-1 using `pg-boss`, create `tasks/2-1-pg-boss-guide.md`

#### âš™ï¸ Task Granularity
- âœ… Tasks should be **as small as practicable**
- âœ… Must still represent a cohesive, testable unit
- âœ… Break down complex features into multiple smaller tasks

#### ğŸ” Don't Repeat Yourself (DRY)
Information should exist in **ONE** location:

- âœ… Task details â†’ in dedicated task files
- âœ… PBI documents â†’ reference tasks, don't duplicate
- âœ… **Exception:** Titles/names can be repeated for clarity

#### ğŸ”¢ Use Constants for Values
- âŒ **BAD:** `for (let i = 0; i < 10; i++)`
- âœ… **GOOD:** `const NUM_WEBSITES = 10; for (let i = 0; i < NUM_WEBSITES; i++)`

**Rule:** Any value used more than once â†’ define as named constant

#### ğŸ“– Technical Documentation
For any PBI creating/modifying APIs or interfaces, create documentation including:

- API usage examples and patterns
- Interface contracts and expected behaviors
- Integration guidelines
- Configuration options and defaults
- Error handling and troubleshooting

**Location:** `docs/technical/` or inline code documentation

#### ğŸ” Principle of Hierarchical Planning (PBI vs. Task)
To maintain DRY, planning information must be hierarchical:

- PBI Detail Document (prd.md): The Implementation Plan and Testing Strategy sections define the high-level, multi-task strategy (e.g., phases, milestones, testing approach).
- Task Detail Document (<ID>-<TaskNum>.md): The Implementation Plan and Test Plan sections define the detailed, step-by-step, executable actions required to complete one part of the PBI's high-level plan.
- The AI_Agent is responsible for decomposing the high-level PBI plan into detailed, actionable Tasks.

#### â™»ï¸ Legacy Code Prioritization
- âœ… When analyzing a PBI or Task, the **AI_Agent** must *always* prioritize searching for and **leveraging** existing functions, APIs, components, and patterns within the codebase.
- âŒ **PROHIBITED:** Writing new code for functionality that already exists.
- âœ… All implementation plans (`Implementation Plan` / `Implementation Steps`) must explicitly state which legacy assets will be reused.
---

### 2.2 AI Agent Automation Rules

**Purpose:** Defines how AI agents interact with workflow system to ensure synchronization and integrity across all tasks and PBIs.

#### Workflow Interaction Principles

The AI_Agent **MUST** automatically ensure:

1. **Automatic Synchronization**
   - âœ… Status consistency between task files and index files
   - âœ… All file updates occur atomically (same commit)
   - âœ… History logs maintained in both locations
   - âœ… No orphaned tasks or PBIs

2. **Workflow State Management**
   - âœ… Follow defined state transitions (see Section 3.3 for PBI, Section 4.3 for Tasks)
   - âœ… Validate pre-conditions before state changes
   - âœ… Execute all required actions during transitions
   - âœ… Log all state changes with timestamp and user

3. **Integrity Verification**
   - âœ… Before starting any task: verify it exists in both task file AND index
   - âœ… Before status change: validate transition is allowed per workflow rules
   - âœ… After any update: verify both locations are synchronized
   - âœ… Detect and report any inconsistencies immediately

#### Task Management Automation Flow

When working with tasks, AI_Agent **AUTOMATICALLY**:

**1. Task Initiation:**
```
User approves task (Proposed â†’ Agreed)
    â†“
AI_Agent MUST:
    â”œâ”€ Create task file: <PBI-ID>-<TASK-ID>.md
    â”œâ”€ Link in tasks index
    â”œâ”€ Populate all required sections
    â”œâ”€ Update status in BOTH locations
    â””â”€ Log creation in history
```

**2. Task Execution:**
```
Task status: Agreed â†’ InProgress
    â†“
AI_Agent MUST verify:
    â”œâ”€ No other InProgress tasks for same PBI
    â”œâ”€ Task status matches in both locations
    â””â”€ All pre-conditions met
    â†“
AI_Agent MUST update:
    â”œâ”€ Task file status
    â”œâ”€ Index file status
    â”œâ”€ Status history in task file
    â””â”€ Timestamp and user info
```

**3. Task Completion:**
```
Task status: InProgress â†’ Review
    â†“
AI_Agent MUST verify:
    â”œâ”€ All requirements implemented
    â”œâ”€ All tests passing
    â””â”€ Files modified documented
    â†“
AI_Agent MUST update:
    â”œâ”€ Task file with implementation details
    â”œâ”€ Task file status â†’ Review
    â”œâ”€ Index file status â†’ Review
    â””â”€ Status history logged
    â†“
User approves (Review â†’ Done)
    â†“
AI_Agent MUST:
    â”œâ”€ Review next task relevance
    â”œâ”€ Confirm with User if next tasks still valid
    â”œâ”€ Update both locations to Done
    â”œâ”€ Execute version control workflow (Section 4.8)
    â””â”€ Log completion
```

#### PBI Management Automation Flow

When working with PBIs, AI_Agent **AUTOMATICALLY**:

**1. PBI Creation:**
```
User creates PBI (â†’ Proposed)
    â†“
AI_Agent MUST verify:
    â”œâ”€ Unique PBI ID
    â”œâ”€ Required fields complete
    â””â”€ PRD alignment (if PRD exists)
    â†“
AI_Agent MUST log:
    â””â”€ Creation in PBI history
```

**2. PBI Plan Generation (User approves idea):**
```
PBI status: Proposed â†’ Agreed (User trigger)
    â†“
AI_Agent MUST:
    â”œâ”€ **(NEW) Scan codebase for reusable assets (Discovery)**
    â”œâ”€ Create PBI detail document: <PBI-ID>/prd.md (using Template 2)
    â”œâ”€ Populate all sections (high-level plan) **using Discovery findings**
    â”œâ”€ Link backlog â†” detail document
    â”œâ”€ Update PBI status â†’ PlanInReview
    â””â”€ Log action in history
```

**3. PBIPBI Task Decomposition (User approves plan):**
```
PBI status: PlanInReview â†’ ReadyForTasks (User trigger)
    â†“
AI_Agent MUST:
    â”œâ”€ Read approved prd.md
    â”œâ”€ Create task list: <PBI-ID>/tasks.md
    â”œâ”€ Decompose plan into all task files (<ID>-<TaskNum>.md)
    â”œâ”€ Set all new tasks to Proposed
    â”œâ”€ Update PBI status â†’ InProgress
    â””â”€ Log action in history
    â””â”€ Begin task execution flow
```

**4. PBI Review (AI finishes all tasks):**
```
PBI status: InProgress â†’ InReview
    â†“
AI_Agent MUST verify:
    â”œâ”€ ALL tasks are Done
    â”œâ”€ All CoS met
    â””â”€ Documentation updated
    â†“
AI_Agent MUST:
    â”œâ”€ Update PBI status
    â”œâ”€ Notify for review
    â””â”€ Log submission
```

#### Automatic Validation Checkpoints

AI_Agent performs these checks **AUTOMATICALLY** at each interaction:

**Before ANY code change:**
```
1. âœ… Identify linked PBI and Task
2. âœ… Verify task is in correct status (InProgress)
3. âœ… Confirm change is within task scope
4. âœ… Check task status matches in both locations
```

**Before ANY status change:**
```
1. âœ… Validate transition is allowed per workflow
2. âœ… Verify all pre-conditions met
3. âœ… Prepare to update both locations atomically
```

**After ANY file update:**
```
1. âœ… Verify synchronization between locations
2. âœ… Confirm history logged
3. âœ… Validate all links still work
```

#### Error Detection and Recovery

AI_Agent **MUST** detect and report:

âŒ **Synchronization Errors:**
- Task status mismatch between file and index
- Orphaned task files (in folder but not in index)
- Missing task files (in index but file doesn't exist)

âŒ **Workflow Violations:**
- Multiple InProgress tasks for same PBI
- Invalid state transitions
- Missing required documentation

âŒ **Integrity Issues:**
- Broken links between documents
- Missing required sections
- Incomplete history logs

**Recovery Actions:**
```
Error detected
    â†“
1. â›” STOP all work immediately
2. ğŸ“¢ Report error to User with details:
   â”œâ”€ What is inconsistent
   â”œâ”€ Where the mismatch exists
   â””â”€ Recommended fix
3. â¸ï¸ Wait for User guidance
4. âœ… Apply fix to both locations atomically
5. âœ… Verify resolution
```

#### Automation Safeguards

To ensure proper workflow compliance, AI_Agent:

âœ… **NEVER:**
- Makes status changes without updating both locations
- Creates tasks without corresponding files
- Modifies files outside current task scope
- Proceeds with work if validation fails

âœ… **ALWAYS:**
- Validates before acting
- Updates atomically
- Logs all changes
- Reports inconsistencies immediately
- Asks User when uncertain

---

### 2.3 PRD Alignment Check

âœ… **MUST** check all PBIs for alignment with PRD
âš ï¸ **MUST** raise any discrepancies with User

---

### 2.4 Integrity and Sense Checking

âœ… All data must be sense-checked for consistency and accuracy

---

### 2.5 Scope Limitations

âŒ **NO gold plating** or scope creep
âœ… All work scoped to specific task
âœ… Improvements â†’ propose as separate tasks

---

### 2.6 Change Management Rules

**Before ANY code change:**

1. âœ… Identify the linked PBI or Task
2. âœ… Verify change is within task scope
3. âœ… If User requests change without task reference:
   - âŒ **DO NOT** proceed with work
   - âœ… Discuss whether to associate with existing task or create new PBI + task

**Change Workflow:**

```
User Request â†’ Identify Task â†’ Verify Scope â†’ Execute OR Create New Task
```

---

## Navigation

- [â† Back to Index](../project-policy-index.md)
- [Next: Section 2 - PBI Management â†’](./2-pbi-management.md)

---

**End of Section 1**
